core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -423,6 +453,9 @@ struct Inner<Block: BlockT> {
 	live_topics: KeepTopics<Block>,
 	config: crate::Config,
 	next_rebroadcast: Instant,
+	incoming_votes_tally: HashMap<Round, HashMap<AuthorityId, VoteTally>>,
+	incoming_msg_tally: HashMap<Round, HashMap<(AuthorityId, PeerId), VoteTally>>,
+	outgoing_votes_tally: HashMap<Round, HashMap<(AuthorityId, PeerId), VoteTally>>,
```

Comment:
Keying these only by round is a little fragile. If we have 2 set changes shortly after another, we could have 2 rounds with the same number.

Reply:
Yes that should be possible, since the set id is available at each call site I think.